core/lib/zksync_core/src/native_erc20_fetcher/mod.rs

```diff
@@ -0,0 +1,115 @@
+use std::{sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use metrics::atomics::AtomicU64;
+use tokio::{
+    sync::{watch, OnceCell},
+    task::JoinHandle,
+};
+use zksync_config::configs::native_erc20_fetcher::NativeErc20FetcherConfig;
+
+// TODO: this error type is also defined by the gasAdjuster module,
+//       we should consider moving it to a common place
+#[derive(thiserror::Error, Debug, Clone)]
+#[error(transparent)]
+pub struct Error(Arc<anyhow::Error>);
+
+impl From<anyhow::Error> for Error {
+    fn from(err: anyhow::Error) -> Self {
+        Self(Arc::new(err))
+    }
+}
+
+pub trait Erc20Fetcher: 'static + std::fmt::Debug + Send + Sync {
+    fn conversion_rate(&self) -> anyhow::Result<u64>;
+}
+
+pub(crate) struct NativeErc20FetcherSingleton {
+    native_erc20_fetcher_config: NativeErc20FetcherConfig,
+    singleton: OnceCell<Result<Arc<NativeErc20Fetcher>, Error>>,
+}
+
+impl NativeErc20FetcherSingleton {
+    pub fn new(native_erc20_fetcher_config: NativeErc20FetcherConfig) -> Self {
+        Self {
+            native_erc20_fetcher_config,
+            singleton: OnceCell::new(),
+        }
+    }
+
+    pub async fn get_or_init(&mut self) -> Result<Arc<NativeErc20Fetcher>, Error> {
+        let adjuster = self
+            .singleton
+            .get_or_init(|| async {
+                let fetcher =
+                    NativeErc20Fetcher::new(self.native_erc20_fetcher_config.clone()).await;
+                Ok(Arc::new(fetcher))
+            })
+            .await;
+        adjuster.clone()
+    }
+
+    pub fn run_if_initialized(
+        self,
+        stop_signal: watch::Receiver<bool>,
+    ) -> Option<JoinHandle<anyhow::Result<()>>> {
+        let fetcher = self.singleton.get()?.clone();
+        Some(tokio::spawn(async move { fetcher?.run(stop_signal).await }))
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct NativeErc20Fetcher {
+    // TODO: we probably need to add a http client here
+    // to avoid creating a new one for each request
+    pub config: NativeErc20FetcherConfig,
+    pub latest_to_eth_conversion_rate: AtomicU64,
+}
+
+impl NativeErc20Fetcher {
+    pub(crate) async fn new(config: NativeErc20FetcherConfig) -> Self {
+        let conversion_rate = reqwest::get(format!("{}/conversion_rate", config.host))
+            .await
+            .unwrap()
```

Comment:
I think it makes sense to change the signature of this method to return `anyhow::Result<Self>` and propagate errors instead of panicking (bonus points for using `anyhow::Context` to add context to these errors 🙃 ).

Reply:
Good idea, tracked [here](https://github.com/lambdaclass/zksync-era/issues/195) (same for [this comment](https://github.com/matter-labs/zksync-era/pull/996#discussion_r1503677702))